### PR TITLE
#28631: Test celu with exp21f

### LINKED ISSUE #28631: Improve celu accuracy
To check and update if celu accuracy can be improved by using exp21f : https://github.com/tenstorrent/tt-metal/pull/25946

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Ticket
#26993, #28631

### Problem description
Test celu with exp21f to see if there is improvement in accuracy

### What's changed
- Updated celu to use exp21
- **ULP Data** 
  - Tested for the entire 16-bit range of bfloat16 (0 through 2¹⁶), covering all possible values. Test passes with ULP 1 except for below mentioned cases
  - Attached ULP Graphs, TTNN vs Torch, Values causing higher ULP : [File](https://docs.google.com/document/d/1jSSbH0YpiJtgrWaiOrpEkK_lmz5TyilARTx8QhAnJFA/edit?tab=t.0#heading=h.fhq5x9wxrs1n)
  - Below listed are the failing cases with ULP>1
    - Main branch : 
        - Max ULP Delta: **128.0** (Mismatching values : [Link](https://drive.google.com/file/d/1KBL0hbw2LaARXIr8-Sb2yz9vGhm2Rk_O/view?usp=drive_link) ) 
        - Input Failing range : (**-4.46875**, 1.1663108012064884e-38 )
        - All close testing
          - (-4.46875, 1.1663108012064884e-38 ) : Max ATOL Delta: 0.01171875
          - (-0.28515625, 1.1663108012064884e-38 ) : Max ATOL Delta: 0.01171875
    - Updated branch : 
        - Max ULP Delta: **1.8775345440461937e+37** (Mismatching value : [Link](https://drive.google.com/file/d/1i_nSKk0A5X8LIPCsifwyuW7vYDbnHoRe/view?usp=drive_link))
        - Input Failing range : (**-0.28515625**, 1.1663108012064884e-38 )
        - All close testing
          - (-0.28515625, 1.1663108012064884e-38 ) : ATOL : 0.001953125
          - BF16 range (-1.6 * 10**38, 1.6 * 10**38) : ATOL : 0.0
- **Edge case behaviour**

<google-sheets-html-origin>
**INPUT** | **TORCH RESULT** | **TT RESULT (Main)** | **TT RESULT (Updated)**
-- | -- | -- | --
inf | inf | inf | inf
-inf | -1 | -1 | -1
nan | nan | inf | inf
-0.0 | 0 | 0 | 0.0017
0 | 0 | 0 | 0

- **Performance analysis**
  - Single tile Performance : Updated kernel duration is about **22%** slower
    - Main Branch : [celu_main_perf.csv](https://github.com/user-attachments/files/22320498/celu_main_perf.csv)
    - Updated branch : [celu_updated_perf.csv](https://github.com/user-attachments/files/22320501/celu_updated_perf.csv)
    - Image
  - 8K tile Performance : Updated kernel duration is about **36.6%** slower
    - Main Branch : [celu_main_8K_perf.csv](https://github.com/user-attachments/files/22321047/celu_main_8K_perf.csv)
    - Updated branch : [celu_updated_8K_perf.csv](https://github.com/user-attachments/files/22321048/celu_updated_8K_perf.csv)
    - Image

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17896758942) - passed
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17896761383) - passed
- [ ] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/17896775338) - failed as in main
- [x] [(Blackhole) Blackhole nightly tests](https://github.com/tenstorrent/tt-metal/actions/runs/17896772335) - passed 
- [ ] [BH PC Watcher enabled](https://github.com/tenstorrent/tt-metal/actions/runs/17896760470) - passes as in main
- [ ] [Single card pipeline](https://github.com/tenstorrent/tt-metal/actions/runs/17896774228) - passes as in main
- [ ] [T3K Pipeline](https://github.com/tenstorrent/tt-metal/actions/runs/17896776907) - passes as in main
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/eltwise/test_celu_21f.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    assert_with_ulp(golden, result, 1, allow_nonfinite=True)
+
+
+@pytest.mark.parametrize(
+    "low, high, expected_Atol, expected_rtol",
+    [
+        (-0.28515625, 1.1663108012064884e-38, 0.002, 0.02),
+        (-1.6 * 10**38, 1.6 * 10**38, 0, 0),  # bf16 range
+    ],
+)
```
Comment (nmauriceTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_celu_21f.py:44]: Can you split the ranges so that they do not overlap ? (same as expm1)

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_celu.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    for (int d = 0; d < ITERATIONS; d++) {
+        sfpi::vFloat v = sfpi::dst_reg[0];
+
+        v_if(v < 0.0f) {
+            // Compute exp(x / alpha)
+            sfpi::vFloat exp_val =
+                _sfpu_exp_21f_<true>(v * alpha_recip);  // is_fp32_dest_acc_en set to true to avoid rounding as it has
+                                                        // to be done at the end of operation
+
+            sfpi::vFloat result = alpha * (exp_val - 1.0f);
```
Comment (nmauriceTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_celu.h]: You should replace 0.0f and 1.0f with `sfpi::vConst0` and `sfpi::vConst1` ?  (same for wormhole version)

File: tests/ttnn/unit_tests/operations/eltwise/test_celu_21f.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+def test_celu_allclose(low, high, expected_Atol, expected_rtol, device):
+    num_elements = torch.prod(torch.tensor(torch.Size([1, 3, 320, 320]))).item()
```
Comment (nmauriceTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_celu_21f.py]: I think you should use:
```
shape = [1, 3, 320, 320]
num_elements = math.prod(shape)
```
This avoids allocating a new tensor and it should be more clear.

**Review by nmauriceTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_celu_21f.py (hunk 1 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/eltwise/test_celu_21f.py:44]: Done

**Review by dchenTT — APPROVED** [review]

**Review by nvelickovicTT — APPROVED** [review]
